File: src/main/java/at/hrechny/predictionsbot/config/MessageResolver.java
package at.hrechny.predictionsbot.config;

import jakarta.inject.Singleton;
import java.text.MessageFormat;
import java.util.Locale;
import java.util.MissingResourceException;
import java.util.ResourceBundle;

@Singleton
public class MessageResolver {

  public String getMessage(String code, Object[] args, Locale locale) {
    var resolvedLocale = locale != null ? locale : Locale.forLanguageTag("ru");
    String pattern;
    try {
      pattern = ResourceBundle.getBundle("messages", resolvedLocale).getString(code);
    } catch (MissingResourceException ex) {
      pattern = code;
    }

    if (args == null || args.length == 0) {
      return pattern;
    }
    return new MessageFormat(pattern, resolvedLocale).format(args);
  }

  public LocalizedMessages forLocale(Locale locale) {
    return new LocalizedMessages(this, locale);
  }
}

File: src/main/kotlin/at/hrechny/predictionsbot/config/MessageResolver.kt
package at.hrechny.predictionsbot.config

import jakarta.inject.Singleton
import java.text.MessageFormat
import java.util.Locale
import java.util.MissingResourceException
import java.util.ResourceBundle

@Singleton
class MessageResolver {
    fun getMessage(code: String, args: Array<Any>?, locale: Locale?): String {
        val resolvedLocale = locale ?: Locale.forLanguageTag("ru")
        val pattern = try {
            ResourceBundle.getBundle("messages", resolvedLocale).getString(code)
        } catch (exception: MissingResourceException) {
            code
        }

        if (args == null || args.isEmpty()) {
            return pattern
        }
        return MessageFormat(pattern, resolvedLocale).format(args)
    }

    fun forLocale(locale: Locale?): LocalizedMessages = LocalizedMessages(this, locale)
}
